asmrun/signals_asm.c

```diff
@@ -214,38 +214,69 @@ DECLARE_SIGNAL_HANDLER(segv_handler)
   struct rlimit limit;
   struct sigaction act;
   char * fault_addr;
+  int handled = 0;
 
   /* Sanity checks:
      - faulting address is word-aligned
      - faulting address is within the stack
      - we are in OCaml code */
   fault_addr = CONTEXT_FAULTING_ADDRESS;
   if (((uintnat) fault_addr & (sizeof(intnat) - 1)) == 0
-      && getrlimit(RLIMIT_STACK, &limit) == 0
-      && fault_addr < system_stack_top
-      && fault_addr >= system_stack_top - limit.rlim_cur - EXTRA_STACK
+      && fault_addr < caml_top_of_stack) {
+    size_t stack_size;
+
+    if (caml_stack_size < 0) {
+      /* A thread (not the main thread) whose stack size couldn't be
+         computed. */
+      stack_size = 0;
+    } else if (caml_stack_size == 0) {
+      /* The current stack is that of the main thread.
+         [getrlimit] is not async signal safe and thus should not be called
+         here, but it isn't clear what else is better.  (In particular the
+         pthread calls that might be used instead are not known to be
+         async signal safe either.)  The stack size for the main thread
+         cannot really be cached as it might grow. */
+      if (getrlimit(RLIMIT_STACK, &limit) == 0) {
+        stack_size = limit.rlim_cur + EXTRA_STACK_RLIMIT;
+      }
+      else {
```

Comment:
`} <newline> else {` is not idiomatic for this code base.  Please write `} else {` like everywhere else.